lib/runners/processor/pmd_cpd.rb

```diff
@@ -0,0 +1,191 @@
+module Runners
+  class Processor::PmdCpd < Processor
+    include Java
+
+    Schema = StrongJSON.new do
+      let :runner_config, Schema::BaseConfig.base.update_fields { |fields|
+        fields.merge!({
+                    'minimum-tokens': numeric?,
+                    files: enum?(string, array(string)),
+                    language: string?,
+                    encoding: string?,
+                    'skip-duplicate-files': boolean?,
+                    'non-recursive': boolean?,
+                    'skip-lexical-errors': boolean?,
+                    'ignore-annotations': boolean?,
+                    'ignore-identifiers': boolean?,
+                    'ignore-literals': boolean?,
+                    'ignore-usings': boolean?,
+                    'no-skip-blocks': boolean?,
+                    'skip-blocks-pattern': string?,
+                  })
+      }
+
+      let :issue, object(
+        lines: integer,
+        tokens: integer,
+        files: array(object(
+          id: string,
+          path: string,
+          start_line: integer,
+          start_column: integer,
+          end_line: integer,
+          end_column: integer
+        )),
+        codefragment: string
+      )
+    end
+
+    DEFAULT_MINIMUM_TOKENS = 100
+    DEFAULT_FILES = ".".freeze
+
+    register_config_schema(name: :pmd_cpd, schema: Schema.runner_config)
+
+    def analyzer_version
+      @analyzer_version ||= capture3!("show_pmd_version").yield_self { |stdout,| stdout.strip }
+    end
+
+    def analyze(changes)
+      run_analyzer
+    end
+
+    def run_analyzer
+      stdout, stderr, status = capture3(analyzer_bin, *cli_options)
+      if status.success? || status.exitstatus == 4
+        Results::Success.new(guid: guid, analyzer: analyzer).tap do |result|
+          construct_result(result, stdout, stderr)
+        end
+      else
+        Results::Failure.new(guid: guid, analyzer: analyzer, message: "Unexpected error occurred. Please see the analysis log.")
+      end
+    end
+
+    def construct_result(result, stdout, stderr)
+      stdout.sub!(/encoding=".*?"/,'') # Remove the encoding attribute to read this XML as UTF-8. The PMD CPD writes an XML report as UTF-8 with the inconsistent encoding attribute value when the --encoding option is specified.
```

Comment:
[suggestion] What about the following refactoring?

- Make the trailing long line comment readable to show it without page scrolling.
- Clarity replacing the XML declaration.

```suggestion
      # HACK: Remove the encoding attribute to read this XML as UTF-8.
      #       The PMD CPD writes an XML report as UTF-8 with the inconsistent encoding attribute value when the --encoding option is specified.
      stdout.sub!(/<?xml version="1.0" encoding=".+"?>/, '<?xml version="1.0"?>')

```

Reply:
I am concerned about changes of the XML declaration in the future, e.g. adding new attributes. But we already have smoke tests to detect the changes, so I have accepted your suggestion.
commit ddaa70724b929a934c6cdff64e86eab062e8f8b5

Also I have replaced `encoding="foo"` with `encoding="UTF-8"` rather than just deleting, so that the parser uses 'UTF-8' instead of the default encoding.
https://github.com/sider/runners/commit/ddaa70724b929a934c6cdff64e86eab062e8f8b5#diff-78e7c4c97abb54a5880bb4582b2322c8R66

Reply:
Thanks!